portality/bll/doaj.py

```diff
@@ -0,0 +1,620 @@
+from portality import constants
+from portality import lock
+from portality import models
+from portality.bll import exceptions
+from portality.core import app
+from portality.lib import dates
+from portality.lib.argvalidate import argvalidate
+
+
+class DOAJ(object):
```

Comment:
I've explicitly avoided thinking about how we'll compose the full business logic layer in this project.  We talked about it at the start of the work, and I proposed that we'd put everything in together right now, and then when we do the refactor of this layer properly we'll agree a structure.  I stand by that - I don't think it's the right time to define the shape of this layer before we've better understood what will be in it.  I therefore don't plan to make any changes here unless there are actual bugs.

With regard to the shape of it, I have been doing some research, and there are a couple of fairly well-trodden options that I think we can go down.  My favourite so far is to separate the BLL into "services", and have each of those services retrievable through a single interface (probably `DOAJ`).  We'll make a judgement call at some point as to how to divide the services up, so that functions commonly used together are near each other in the imports, just to minimise the number of imports necessary for any task (this will be a hand-wavy exercise, there's not a huge gain to be had, it will be mostly for code comprehensibility).

Each of those services then contains all the functions that you can execute in that area.

So, for example:

```python
journalService = DOAJ.journalService()
journalService.journal_2_application()
```

This gives us some leeway in the future for modifying behaviour at run time (e.g for testing), since DOAJ then becomes a Factory, which can decide which service to give us.

I recommend against the following approaches:

* Composition - it makes `DOAJ()` difficult to maintain, and risks name clashes in functions, plus requires a full import of everything
* Putting everything in `DOAJ()` - the class would just be too large to reasonably work on, and again everything would need to get loaded every time

Reply:
Changes to the business layer in line with our technical meeting have now been done.